### PR TITLE
Fixed a bug calculating cross entropy loss in `JetMoeForCausalLM`

### PR DESCRIPTION
**Pull request by Phoenix-Shen**
In the original code, we shift the logits and pass `shift_logits` into the `self.loss_function`, but in `self.loss_function`, the `shift_logits` will be shifted again, so we are actually doing **next next token prediction**, which is incorrect. I have removed the logits shifting before calling `self.loss_function`.

# What does this PR do?

Fixed incorrect cross entropy calculation in `JetMoeForCausalLM`. 

The original code snippet of the model's forward function is:
```python
if labels is not None:

  logits = logits.float()
  # NOTICE: The labels are shifted once here
  shift_logits = logits[..., :-1, :].contiguous()
  shift_labels = labels[..., 1:].contiguous()

  shift_logits = shift_logits.view(-1, self.config.vocab_size)
  shift_labels = shift_labels.view(-1)

  shift_labels = shift_labels.to(shift_logits.device)
  loss = self.loss_function(
      shift_logits,
      shift_labels,
      vocab_size=self.config.vocab_size,
      **kwargs,
  )
```
but in `self.loss_function`, typically `ForCausalLMLoss`:
```python
def ForCausalLMLoss(
    logits,
    labels,
    vocab_size: int,
    num_items_in_batch: Optional[int] = None,
    ignore_index: int = -100,
    shift_labels: Optional[torch.Tensor] = None,
    **kwargs,
) -> torch.Tensor:
    logits = logits.float()
   # NOTICE: In the above call, the shift_labels is None (we pass shift_labels as labels, not shift_labels), so the labels will be shifted again!
    if shift_labels is None:
        labels = nn.functional.pad(labels, (0, 1), value=ignore_index)
        shift_labels = labels[..., 1:].contiguous()

    logits = logits.view(-1, vocab_size)
    shift_labels = shift_labels.view(-1)

    shift_labels = shift_labels.to(logits.device)
    loss = fixed_cross_entropy(logits, shift_labels, num_items_in_batch, ignore_index, **kwargs)
    return loss
```

So we are doing **next next token prediction**, which is incorrect. We need to remove the first label shifting operation to fix this bug, that's what I do in this PR. With the modified code snippet, I can get reasonable results in SFT tasks.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@ArthurZucker

### CODE REVIEW

**Comment by Phoenix-Shen** [comment]
Here are the details about how I find this bug, it is interesting:
- I am fine-tuning this model with commonsense reasoning dataset (15K samples)
- The expected output is: `the correct answer is answer1`
- But the model outputs: ` the answer answer1` (missing `correct` and `is`)
- So I suspect the problem could be the loss calculation: the model is doing *next next token* prediction rather than next token prediction, maybe the labels are shifted twice?
- After checking the loss calculation function, I finally locate the bug.

**Review by ArthurZucker — APPROVED** [review]
Happy to merge!
